timers/timers.go

```diff
@@ -0,0 +1,245 @@
+}
+
+func newTimer(deadline uint64, f func()) *Timer {
+	t := &Timer{
+		f:        f,
+		deadline: deadline,
+	}
+	t.tail = t
+	return t
+}
+
+// Stop cancels the deferred operation. It returns whether or not the
+// cancellation succeeded.
+func (t *Timer) Stop() bool {
+	return t.state.CAS(stateScheduled, stateCanceled)
+}
+
+func (t *Timer) expire() bool {
+	return t.state.CAS(stateScheduled, stateExpired)
+}
+
+func (t *Timer) pushOne(head *Timer) *Timer {
+	head.next = t
+	if t == nil {
+		head.tail = head
+	} else {
+		head.tail = t.tail
+		t.tail = nil
+	}
+	return head
+}
+
+func (t *Timer) push(head *Timer) *Timer {
+	if head == nil {
+		return t
+	}
+	if t == nil {
+		return head
+	}
+	head.tail.next = t
+	head.tail = t.tail
+	t.tail = nil
+	return head
+}
+
+type bucket struct {
+	sync.Mutex
+	timers *Timer
+	// Avoid false sharing:
+	// http://mechanical-sympathy.blogspot.com/2011/07/false-sharing.html
+	_ [cacheline - unsafe.Sizeof(sync.Mutex{}) - unsafe.Sizeof(&Timer{})]byte
+}
+
+type bucketList struct {
+	mask    uint64
+	buckets []bucket
+}
+
+func (bs *bucketList) Schedule(deadline uint64, f func()) *Timer {
+	t := newTimer(deadline, f)
+	b := &bs.buckets[deadline&bs.mask]
+	b.Lock()
+	b.timers = b.timers.pushOne(t)
+	b.Unlock()
+	return t
+}
+
+func (bs *bucketList) Clear() {
+	for i := range bs.buckets {
+		b := &bs.buckets[i]
+		b.Lock()
+		b.timers = nil
+		b.Unlock()
+	}
+}
+
+func (bs *bucketList) GatherExpired(start, end, now uint64) *Timer {
+	// If we're more than a full rotation behind, only inspect each bucket
+	// once.
+	if (end-start == math.MaxUint64) || (int(end-start+1) > len(bs.buckets)) {
+		start, end = 0, uint64(len(bs.buckets))
+	}
+
+	var todo *Timer
+	for tick := start; tick < end; tick++ {
+		todo = bs.gatherBucket(tick, now, todo)
+	}
+	return todo
+}
+
+func (bs *bucketList) gatherBucket(tick, now uint64, todo *Timer) *Timer {
+	b := &bs.buckets[tick&bs.mask]
+
+	b.Lock()
+	batch := b.timers
+	b.timers = nil
+	b.Unlock()
+
+	var unexpired *Timer
+	for batch != nil {
+		next := batch.next
+		if batch.deadline > now {
+			unexpired = unexpired.pushOne(batch)
+		} else if batch.expire() {
+			todo = todo.pushOne(batch)
+		}
+		batch = next
+	}
+	if unexpired != nil {
+		// We should only hit this case if we're a full wheel rotation
+		// behind.
+		b.Lock()
+		b.timers = b.timers.push(unexpired)
+		b.Unlock()
+	}
+	return todo
+}
+
+// A Wheel schedules and executes deferred operations.
+type Wheel struct {
+	clock     clock.Clock
+	periodExp uint64
+	ticker    *clock.Ticker
+	buckets   bucketList
+	stopOnce  sync.Once
+	stop      chan struct{}
+	stopped   chan struct{}
+}
+
+// NewWheel creates and starts a new Wheel.
+func NewWheel(period, maxTimeout time.Duration) *Wheel {
+	w := newWheel(period, maxTimeout, clock.New())
+	w.start()
+	return w
+}
+
+func newWheel(period, maxTimeout time.Duration, clock clock.Clock) *Wheel {
+	tickNanos, power := nextPowerOfTwo(int64(period))
```

Comment:
For `maxTimeout`, we round up to be be conservative and give use a higher max than what was specified so we can just mask. For period, I think the conservative option would be to round down right? If the user asks for 5ms, using a tick period of ~4ms would be better than what the user asked for.

I think that can be done easily by doing:
```go
tickNanos, power := nextPowerOfTwo(int64(period)/2+1)
```

If the user specifies 5ms, we would find the next power of 2 after 2.5ms, which is 4ms. If the user specifies an exact power of 2 like 4.194..ms,  then we divide by 2 but then add 1 so we don't select ~2ms.